### ISSUE TITLE
Intents component: fired intents behave different in Studio and ingame

### ISSUE THREAD

**Issue by DireMitten** [CrescentCode/Freya#3]
Tl;dr: Firing an intent from the client in Studio doesn't automatically pass the local player as an argument. When testing ingame, it behaves as expected (i.e. IsLocal and the player who fired the intent is automatically passed before any of the other arguments)

I'm using an intent in a control script for the client to ask to fire the gun on their tank. This is what it looks like on the client:

Intent:Fire("TankFired", Player.Character)

I'm passing in a single argument, the character of the player belonging to the client. This is what it looks like on the server:

function tankFired(IsLocal, player, tank, shouldBeNil)
    print(isLocal, player, tank, shouldBeNil)
    if tank == player.Character then
        --shoot bullets and stuff
    end
end

Intent:Subscribe("TankFired", tankFired)

When I test it ingame by pressing play on the Roblox page, it works as expected. The output is

false direMitten Blue Tank nil

Which makes perfect sense. But when I test it in Studio (the blue Play button), I get an error "Character is not a valid member of Model", which makes sense because the output reads

true Blue Tank nil nil

In other words, it doesn't automatically pass the Player as an argument when the intent is fired. This might be intentional on your part, but it's very unwieldy. I could add checks to see if the script is running in Studio, and change the passed arguments based on that, but I don't think I'm being unreasonable when I say that I don't want to spend time on or worry about that.

I understand that this behavior is similar to the built in BindableEvents and RemoteEvents, in that a player argument is obviously not passed when stuff isn't sent between client and server, which is *technically* not the case when testing things in Studio. 

Am I supposed to be using the Tell method instead of the Fire method? I tried, and it works the same ingame. In Studio, nothing happens at all. It seems like the intent doesn't fire at all?

Oh and by the way, testing in Studio with a Local server does behave as expected. This would be fine, if it wasn't for the fact that doing this takes several times longer than just pressing Play.

**Comment by eLunate** [comment]
Unfortunately, due to the way that Roblox doesn't internally separate the local and remote environments in Studio local testing (not a test server), the intent doesn't know that it is on the client (there is no way to get the script type without getfenv and looking for the `script` object.), resulting in the behaviour you're seeing.

Iirc Intents do spit out a warning when you try to load them in Studio play mode, stating that they may not work as expected. If not, I'll add the check and commit it.

Blame Roblox, as usual.
